File: src/co/nums/intellij/aem/htl/psi/impl/HtlPsiUtil.java
package co.nums.intellij.aem.htl.psi.impl;

import co.nums.intellij.aem.htl.psi.HtlExpression;
import com.intellij.lang.html.HTMLLanguage;
import com.intellij.psi.FileViewProvider;
import com.intellij.psi.PsiElement;
import com.intellij.psi.util.PsiTreeUtil;
import com.intellij.psi.xml.XmlAttributeValue;
import org.jetbrains.annotations.Nullable;

public final class HtlPsiUtil {

	private HtlPsiUtil() {
		// no instances
	}

	@Nullable
	public static String getEnclosingHtmlAttributeQuote(PsiElement element) {
		HtlExpression expression = PsiTreeUtil.getParentOfType(element, HtlExpression.class);
		if (expression != null) {
			XmlAttributeValue xmlAttributeValue = getEnclosingXmlAttributeValue(expression);
			if (xmlAttributeValue != null) {
				return Character.toString(xmlAttributeValue.getText().charAt(0));
			}
		}
		return null;
	}

	@Nullable
	private static XmlAttributeValue getEnclosingXmlAttributeValue(PsiElement element) {
		int offset = element.getTextOffset();
		if (offset > 0) {
			FileViewProvider viewProvider = element.getContainingFile().getViewProvider();
			PsiElement previousElement = viewProvider.getPsi(HTMLLanguage.INSTANCE).findElementAt(offset - 1);
			if (previousElement != null) {
				return PsiTreeUtil.getParentOfType(previousElement, XmlAttributeValue.class);
			}
		}
		return null;
	}

}

File: src/co/nums/intellij/aem/htl/psi/impl/HtlPsiUtil.kt
package co.nums.intellij.aem.htl.psi.impl

import co.nums.intellij.aem.htl.psi.HtlExpression
import com.intellij.lang.StdLanguages
import com.intellij.psi.PsiElement
import com.intellij.psi.util.PsiTreeUtil
import com.intellij.psi.xml.XmlAttributeValue

object HtlPsiUtil {

    /**
     * Returns outer HTML attribute quote if HTL expression is inside of
     * attribute.

     * @param element element from expression to check
     * *
     * @return outer HTML attribute quote or `null` if element is not
     * * enclosed by attribute
     */
    fun getOuterHtmlAttributeQuote(element: PsiElement): Char? {
        val htlExpression = PsiTreeUtil.getParentOfType(element, HtlExpression::class.java) ?: return null
        val xmlAttributeValue = getOuterXmlAttributeValue(htlExpression) ?: return null
        return xmlAttributeValue.text.getOrNull(0)
    }

    private fun getOuterXmlAttributeValue(htlExpression: HtlExpression): XmlAttributeValue? {
        val offset = htlExpression.textOffset
        if (offset > 0) {
            val viewProvider = htlExpression.containingFile.viewProvider
            val previousElement = viewProvider.findElementAt(offset - 1, StdLanguages.HTML) ?: return null
            return PsiTreeUtil.getParentOfType(previousElement, XmlAttributeValue::class.java)
        }
        return null
    }

}
